frontend/src/main/java/com/google/tripmeout/frontend/servlet/PlaceParentServlet.java

```diff
@@ -0,0 +1,89 @@
+package com.google.tripmeout.frontend.servlet;
+
+import com.google.common.flogger.FluentLogger;
+import com.google.gson.Gson;
+import com.google.gson.JsonParseException;
+import com.google.tripmeout.frontend.PlaceVisitModel;
+import com.google.tripmeout.frontend.error.TripMeOutException;
+import com.google.tripmeout.frontend.places.PlaceService;
+import com.google.tripmeout.frontend.storage.PlaceVisitStorage;
+import java.io.IOException;
+import java.io.PrintWriter;
+import java.util.List;
+import java.util.UUID;
+import java.util.regex.Pattern;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+public class PlaceParentServlet extends HttpServlet {
+  private static final FluentLogger logger = FluentLogger.forEnclosingClass();
+  private static final Pattern TRIP_NAME_PATTERN = Pattern.compile(".*/trips/([^/]+)/placeVisits");
+
+  private final PlaceVisitStorage placeStorage;
+  private final Gson gson;
+  private final PlaceService placeService;
+
+  public PlaceParentServlet(PlaceVisitStorage placeStorage, Gson gson, PlaceService placeService) {
+    this.placeStorage = placeStorage;
+    this.gson = gson;
+    this.placeService = placeService;
+  }
+
+  @Override
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    PrintWriter writer = response.getWriter();
+    try {
+      PlaceVisitModel place =
+          ServletUtil.extractFromRequestBody(request.getReader(), gson, PlaceVisitModel.class);
+
+      placeService.validatePlaceId(place.placesApiPlaceId());
+
+      String tripId = ServletUtil.matchUriOrThrowError(request, TRIP_NAME_PATTERN).group(1);
+      logger.atInfo().log("finished matching");
+
+      PlaceVisitModel newPlace = PlaceVisitModel.builder()
+                                     .setId(UUID.randomUUID().toString())
+                                     .setTripId(tripId)
+                                     .setPlaceName(place.placeName())
+                                     .setPlacesApiPlaceId(place.placesApiPlaceId())
+                                     .setUserMark(PlaceVisitModel.UserMark.YES)
+                                     .build();
+
+      PlaceVisitModel updatedPlace =
+          placeStorage.updateUserMarkOrAddPlaceVisit(newPlace, PlaceVisitModel.UserMark.YES);
```

Comment:
It's strange to use `updateUserMarkOrAddPlaceVisit` to presumably _insert_ a trip visit for the first time. Even stranger is the fact that you're already setting the user mark above (explicitly to `YES`, as a matter of fact, and ignoring any marks set by the client).

This call should use `addPlaceVisit()`. If we specifically do not want to throw in the case where the place visit already exists, then we should update `PlaceVisitStorage` to provide the semantics we want. In this case, it looks like we want "upsert" (insert-or-update), irrespective of the specific fields that are being changed. If we go that route, we should also remove this update-user-mark method.

Note: we still need to make sure we don't let clients overwrite our server-computed IDs (or other non-client fields).

Reply:
so I was thinking that this method would be called only when a user is adding a placeVisit themselves, in which case I just assumed that meant the user wanted to go (though I guess they could also be adding it for the maybe/if time section)